expandable_segments <-> other allocator options

zdevito opened the pull request:
Previously setting  garbage_collection_threshold or max_split_size_mb along with expandable_segments:True could cause the allocator to hit assert failures when running nearly out of memory. This PR ensures garbage_collection and max_split freeing do not accidentally try to release expandable segments.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #134338

zdevito committed:
expandable_segments <-> other allocator options

[ghstack-poisoned]

zdevito committed:
Update on "expandable_segments <-> other allocator options"




[ghstack-poisoned]

ezyang commented on c10/cuda/CUDACachingAllocator.cpp:
```diff
@@ -2749,7 +2749,8 @@ class DeviceCachingAllocator {
     auto it = pool.blocks.lower_bound(&key);
-    if (it == pool.blocks.end() || (*it)->stream != p.stream()) {
+    if (it == pool.blocks.end() || (*it)->stream != p.stream() ||
+        (*it)->expandable_segment_) {
```
Technically, couldn't there be a block that is big enough, but the biggest block is expandable segment and so we picked out the wrong one in the lower_bound test?

zdevito replied:
Yeah, but in that case, it will just go to the else statement and free that single block anyway.

ezyang reviewed (approved):
This looks good.

zdevito commented:
@pytorchbot rebase

zdevito committed:
Update

[ghstack-poisoned]

zdevito committed:
Update on "expandable_segments <-> other allocator options"

Previously setting  garbage_collection_threshold or max_split_size_mb along with expandable_segments:True could cause the allocator to hit assert failures when running nearly out of memory. This PR ensures garbage_collection and max_split freeing do not accidentally try to release expandable segments.




[ghstack-poisoned]

zdevito commented:
@pytorchbot merge
